test/integration/src/test/java/io/pravega/test/integration/TransactionMetricsTest.java

```diff
@@ -0,0 +1,201 @@
+        serviceBuilder.close();
+        zkTestServer.close();
+    }
+
+    /**
+     * Verify that transaction segment metrics counters are being correctly reported.
+     */
+    @Test(timeout = 20000)
+    public void transactionSegmentMetricsTest() throws TxnFailedException {
+        final String scope = "scope";
+        final String streamName = "stream";
+        final int parallelism = 1;
+        int iterations = 4;
+
+        StreamConfiguration streamConfiguration = StreamConfiguration.builder()
+                .scalingPolicy(ScalingPolicy.fixed(parallelism))
+                .build();
+        @Cleanup
+        StreamManager streamManager = StreamManager.create(controllerURI);
+        streamManager.createScope(scope);
+        streamManager.createStream(scope, streamName, streamConfiguration);
+        @Cleanup
+        ClientFactory clientFactory = ClientFactory.withScope(scope, ClientConfig.builder()
+                .controllerURI(controllerURI)
+                .build());
+        @Cleanup
+        EventStreamWriter<String> writer = clientFactory.createEventWriter(streamName, new JavaSerializer<>(),
+                EventWriterConfig.builder().build());
+
+        int writeBytes = 0;
+        int writeEvents = 0;
+        for (int i = 0; i < iterations; i++) {
+            Transaction<String> transaction = writer.beginTxn();
+
+            // Write some events in the transaction, in this case, of 160 bytes.
+            for (int j = 0; j < 10; j++) {
+                transaction.writeEvent(String.valueOf(j));
+            }
+
+            // Test counters for aborted and committed transaction segments.
+            if (i % 2 == 0) {
+                transaction.commit();
+                writeBytes += 160;
+                writeEvents += 10;
+            } else {
+                transaction.abort();
+            }
+
+            checkCommitOrAbortMetric(getCounter(getSegmentWriteBytesMetricName(scope, streamName, 0)), writeBytes);
+            checkCommitOrAbortMetric(getCounter(getSegmentWriteEventsMetricName(scope, streamName, 0)), writeEvents);
+        }
+    }
+
+    private Counter getCounter(String counterName) {
+        Counter counter = null;
+        // Access the cache until the metric is available.
+        while (counter == null) {
+            counter = MetricRegistryUtils.getCounter(counterName);
+            Exceptions.handleInterrupted(() -> Thread.sleep(100));
+        }
+
+        return counter;
+    }
+
+    private void checkCommitOrAbortMetric(Counter metric, int expectedValue) {
+        boolean updatedCounter = false;
+        do {
+            try {
+                Assert.assertNotNull(metric);
+                Assert.assertEquals(expectedValue, metric.getCount());
+                updatedCounter = true;
+            } catch (AssertionError e) {
+                log.info("Metric not updated in the cache. Retrying.", e);
+                Exceptions.handleInterrupted(() -> Thread.sleep(100));
```

Comment:
Is there no better way of doing this than using a busy-wait loop?

Reply:
You're right, we shouldn't have too many unit integration tests - as of today it already takes over 40 minutes to finish all unit tests.
I've deleted this integration test, and added some mock tests.